enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/group_manager.clj

```diff
@@ -0,0 +1,56 @@
+(ns metabase-enterprise.advanced-permissions.models.permissions.group-manager
+  (:require [clojure.data :as data]
+            [clojure.set :refer [subset?]]
+            [metabase.api.common :as api]
+            [metabase.models :refer [PermissionsGroupMembership]]
+            [metabase.util :as u]
+            [metabase.util.i18n :refer [tru]]
+            [toucan.db :as db]))
+
+(defn user-group-memberships
+  "Return a list of group memberships a User belongs to.
+  Group Membership is a map with 2 keys [:id :is_group_manager]."
+  [user-or-id]
+  (when user-or-id
+    (db/select [PermissionsGroupMembership [:group_id :id] :is_group_manager] :user_id (u/the-id user-or-id))))
+
+(defn- user-group-memberships->map
+  "Transorm user-group-memberships to a map with keys are group-id and values are membership info.
+
+  [{:id 1, :is_group_manager true}] => {1 {:is_group_manager true}}
+
+  We can diff this map to decide which membership to adds/removes."
+  [user-group-memberships]
+  (into {} (map (fn [x] [(:id x) (dissoc x :id)]) user-group-memberships)))
+
+(defn set-user-group-memberships!
+  "Update User Groups Memberships when `advanced-permissions` is enabled.
+  It's used to add/remove users from groups and also promote/demote Group Managers."
```

Comment:
nit: could you make it clear in the doc string that this function only modifies group memberships for a single user? `it's used to add/remove users from groups` made me think it would operate on multiple users